follower/execution/follower.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package execution
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage/badger/operation"
+
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+type blockReader interface {
+	Read(blockID flow.Identifier) ([]byte, error)
+}
+
+type unmarshaler interface {
+	Unmarshal(compressed []byte, value interface{}) error
+}
+
+type Follower struct {
+	log zerolog.Logger
+
+	blocks blockReader
+	codec  unmarshaler
+	db     *badger.DB
+
+	block BlockData
+}
+
+func New(log zerolog.Logger, blocks blockReader, codec unmarshaler, db *badger.DB) *Follower {
+	f := Follower{
+		log: log,
+
+		blocks: blocks,
+		codec:  codec,
+		db:     db,
+	}
+
+	return &f
+}
+
+func (f *Follower) Height() uint64 {
+	if f.block.Block != nil {
+		return f.block.Block.Header.Height
+	}
+	return math.MaxUint64
+}
+
+func (f *Follower) Update() (*ledger.TrieUpdate, error) {
+	if len(f.block.TrieUpdates) == 0 {
+		// No more trie updates are available.
+		return nil, dps.ErrTimeout
+	}
+
+	// Copy next update to be returned.
+	update := f.block.TrieUpdates[0]
+
+	// Move the slice forward by popping the first element.
+	f.block.TrieUpdates = f.block.TrieUpdates[1:]
+
+	return update, nil
+}
+
+// Block returns the latest sealed block data.
+func (f *Follower) Block() BlockData {
+	return f.block
+}
+
+func (f *Follower) OnBlockFinalized(finalizedBlockID flow.Identifier) {
+	for {
+		b, err := f.blocks.Read(finalizedBlockID)
+		if err != nil {
+			// The block data is not yet available. Retry until it becomes available.
+			time.Sleep(1 * time.Second)
+		}
+
+		var block BlockData
+		err = f.codec.Unmarshal(b, &block)
+		if err != nil {
+			f.log.Error().Err(err).Msg("could not unmarshal block from execution follower")
+			return
+		}
+
+		f.block = block
+		break
+	}
+
+	f.IndexAll(finalizedBlockID)
+}
+
+func (f *Follower) IndexAll(blockID flow.Identifier) {
```

Comment:
I'm definitely not on board with this. The follower should not be indexing anything. You are mixing business logic into a component that should be purely about providing data. It also creates significant redundancy where it's truly not necessary. There is no reason not to reuse all of the current transitions logic.

Reply:
I don't get what you mean about the transitions? This simply writes in the same index as the one used by the consensus follower, so that it can then be read by the chain, using the current transitions logic. It does not write in our final index. Maybe this should be clarified.

Reply:
This would slow down the indexing significantly. There is no reason to write anything to disk here, we can keep everything in cache until it has been indexed by the DPS.